File: translated/f-to-c/DRB132-taskdep4-orig-omp45-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//The second taskwait ensures that the second child task has completed; hence it is safe to access
//the y variable in the following print statement.

#include <omp.h>
#include <stdio.h>

void foo()
{
    int x, y;
    x = 0;
    y = 2;

    #pragma omp task depend(inout: x) shared(x)
    x = x + 1;  // 1st Child Task

    #pragma omp task shared(y)
    y = y - 1;  // 2nd child task

    #pragma omp task depend(in: x) if(0)  // 1st taskwait (undeferred)
    {
        // Empty task body - acts as taskwait for x-dependent tasks only
    }

    printf("x = %d\n", x);

    #pragma omp taskwait  // 2nd taskwait - ensures all tasks complete

    printf("y = %d\n", y);  // No race: taskwait ensures completion
}

int main()
{
    #pragma omp parallel
    {
        #pragma omp single
        foo();
    }

    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB132-taskdep4-orig-omp45-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//The second taskwait ensures that the second child task has completed; hence it is safe to access
//the y variable in the following print statement.

package main

import (
	"fmt"
	"sync"
)

func foo() {
	var x, y int
	x = 0
	y = 2

	//$omp task depend(inout: x) shared(x)
	go func() {
		x = x + 1 //1st Child Task
	}()
	//$omp end task

	//$omp task shared(y)
	go func() {
		y = y - 1 //2nd child task
	}()
	//$omp end task

	//$omp task depend(in: x) if(.FALSE.)    //1st taskwait
	//$omp end task

	fmt.Printf("x= %d\n", x)

	//$omp taskwait                          //2nd taskwait

	fmt.Printf("y= %d\n", y)
}

func main() {
	//$omp parallel
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		//$omp single
		foo()
		//$omp end single
	}()
	wg.Wait()
	//$omp end parallel
}
